### PR TITLE
Set response trace context based on that of incoming requests

### PR DESCRIPTION
**Pull request by kjin**
Previously, plugins that patched web frameworks would add in response headers the trace context corresponding to the locally-created root span. This should actually be set to the trace context of the incoming request (modified based on whether this request was traced).

### CODE REVIEW

File: src/plugins/plugin-express.js [hunk 1]
```diff
@@ -40,15 +40,18 @@ function patchModuleRoot(express, api) {
       skipFrames: 3
     };
     api.runInRootSpan(options, function(rootSpan) {
+      // Set response trace context.
+      var outgoingTraceContext =
```
Comment (ofrobots) [inline src/plugins/plugin-express.js]: `responseTraceContext` here and elsewhere.

File: src/plugins/plugin-hapi.js [hunk 2]
```diff
@@ -54,14 +61,11 @@ function createMiddleware(api) {
     
       // we use the path part of the url as the span name and add the full
       // url as a label
-      // req.path would be more desirable but is not set at the time our middlewear runs.
+      // req.path would be more desirable but is not set at the time our middleware runs.
```
Comment (ofrobots) [inline src/plugins/plugin-hapi.js]: lol

File: src/trace-api.js [hunk 3]
```diff
@@ -297,4 +338,4 @@ function createRootSpan_(api, options, skipFrames) {
     incomingTraceContext.spanId,
     skipFrames + 1);
   return new RootSpan(api.agent_, rootContext);
-}
+}
```
Comment (ofrobots) [inline src/trace-api.js]: Add back the newline.

File: src/util.js [hunk 4]
```diff
@@ -70,10 +70,12 @@ function parseContextFromHeader(str) {
       (matches[2] && isNaN(matches[2]))) {
     return null;
   }
+  // We assume that if no value is set for options, we treat the incoming
+  // context as if it had options set to 1.
```
Comment (ofrobots) [inline src/util.js]: Doesn't match my [expectations](https://docs.google.com/document/d/1A4yPsG__X1T4JbZZrT0GcOkCkqRIi48-eY9ZZJUdF-Y/edit).

**Comment by kjin** [comment]
@ofrobots I haven't done any test refactoring in this PR, what do you mean by that the "two commits aren't individually landable"?

**Comment by kjin** [comment]
Oh I see - I was planning to squash these commits.

**Comment by kjin** [comment]
@GoogleCloudPlatform/node-team CI passed, PTAL

File: src/trace-api.js [hunk 5]
```diff
@@ -179,6 +191,31 @@ TraceApiImplementation.prototype.createChildSpan = function(options) {
...
+ * header, the string to be set as this header's value. Otherwise, an empty
+ * string.
+ */
+TraceApiImplementation.prototype.getResponseTraceContext = function(
+    incomingTraceContext, isTraced) {
+  var traceContext = this.agent_.parseContextFromHeader(incomingTraceContext);
+  if (!traceContext) {
+    return '';
+  }
+  traceContext.options = traceContext.options && isTraced ? 1 : 0;
```
Comment (matthewloring) [inline src/trace-api.js]: Not sure which is more readable but `traceContext.options & isTraced` should also work here.

File: src/trace-api.js [hunk 6]
```diff
@@ -221,6 +258,7 @@ var phantomApiImpl = {
   enhancedDatabaseReportingEnabled: function() { return false; },
   runInRootSpan: function(opts, fn) { return fn(null); },
   createChildSpan: function(opts) { return null; },
+  getResponseTraceContext: function(context, traced) { return null; },
```
Comment (matthewloring) [inline src/trace-api.js]: Should this return `''` to keep things stringy?

File: src/util.js [hunk 7]
```diff
@@ -73,7 +73,7 @@ function parseContextFromHeader(str) {
   return {
     traceId: matches[1],
     spanId: matches[2],
-    options: Number(matches[3])
+    options: typeof(matches[3]) === 'undefined' ? 0 : Number(matches[3])
```
Comment (matthewloring) [inline src/util.js]: This is a really sneaky bug!! Could you add a unit test for this as well?

Comment (matthewloring) [inline src/util.js]: Also, maybe `Number(matches[3] || 0)` is more readable? In either case, could you add a comment as well?

Comment (kjin) [inline src/util.js]: Done

**Pull request merged**